File: python/743.network-delay-time.py
import heapq
class Solution(object):
    def solve(self, times, n, k):
        graph = {i:[] for i in range(n)}
        for u,v,w in times:
            graph[u-1].append([v-1, w])
        ##dijkstra...
        dist = n*[1000000]
        dist[k-1] = 0
        pq = [[dist[k-1], k-1]] # dist, start
        while len(pq):
            cur_dist, _from = heapq.heappop(pq)
            if cur_dist > dist[_from]:
                continue

            for _to, w in graph[_from]:
                if dist[_from] + w >= dist[_to]:
                    continue
                dist[_to] = dist[_from]+w
                heapq.heappush(pq, [dist[_to], _to])
        if max(dist) == 1000000:
            return -1
        return max(dist)

    def networkDelayTime(self, times, n, k):
        """
        :type times: List[List[int]]
        :type n: int
        :type k: int
        :rtype: int
        """
        return self.solve(times, n, k)

File: cpp/743.network-delay-time.cpp
class Solution {
public:
    int solve(vector<vector<int> > &times, int n, int k){
	vector<long long> dist(n+1, INT_MAX);
	unordered_map<int, vector<tuple<int, long long>>>adjlist;
	for(auto edge:times){
		adjlist[edge[0]].push_back({edge[1], edge[2]});
	}
	priority_queue<
		   tuple<long long, int>, 
		   vector<tuple<long long, int>>, greater<> >pq;
	dist[k] = 0;
	pq.push({dist[k], k});
	while(!pq.empty()){
	     auto [wu, u] = pq.top(); pq.pop();
	     if(wu > dist[u])continue;
	     for(auto [v, w]:adjlist[u]){
		long long next_dist = dist[u]+w;
		if(next_dist >= dist[v])continue;
		dist[v] = next_dist;
		pq.push({next_dist, v});
	     }
	}
	long long res = INT_MIN;
	for(int i = 1 ; i <= n; i++){
	   res = max(res, dist[i]);
	   if(dist[i]==INT_MAX)return -1;
	}
	return res;
    }
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
	return solve(times, n, k);
    }
};
